File: apps/webflux/firstcome-firstserved/src/main/java/com/booster/firstcomefirstserved/common/error/GlobalExceptionHandler.java
package com.booster.firstcomefirstserved.common.error;

import com.booster.firstcomefirstserved.common.response.ApiResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.bind.support.WebExchangeBindException;
import reactor.core.publisher.Mono;

@Slf4j
@RestControllerAdvice
public class GlobalExceptionHandler {

    // 1. 비즈니스 로직 예외 (우리가 만든 BusinessException)
    @ExceptionHandler(BusinessException.class)
    public Mono<ResponseEntity<ApiResponse<Void>>> handleBusinessException(BusinessException e) {
        log.warn("Business Exception: {}", e.getMessage());
        return Mono.just(
                ResponseEntity.badRequest().body(ApiResponse.fail(e.getErrorCode().getMessage()))
        );
    }

    // 2. 요청 값 검증 실패 (Request DTO의 @NotNull, @Min 등)
    // WebFlux에서는 MethodArgumentNotValidException 대신 WebExchangeBindException이 발생함
    @ExceptionHandler(WebExchangeBindException.class)
    public Mono<ResponseEntity<ApiResponse<Void>>> handleBindException(WebExchangeBindException e) {
        String errorMessage = e.getBindingResult().getAllErrors().get(0).getDefaultMessage();
        log.warn("Validation Failed: {}", errorMessage);
        return Mono.just(
                ResponseEntity
                        .badRequest()
                        .body(ApiResponse.fail(errorMessage))
        );
    }

    // 3. 그 외 알 수 없는 예외
    @ExceptionHandler(Exception.class)
    public Mono<ResponseEntity<ApiResponse<Void>>> handleException(Exception e) {
        log.error("Unhandled Exception: ", e);
        return Mono.just(
                ResponseEntity
                        .internalServerError()
                        .body(ApiResponse.fail("서버 내부 오류가 발생했습니다."))
        );
    }
}

File: apps/kotlin/shopping-service/src/main/kotlin/com/booster/kotlin/shoppingservice/common/exception/GlobalExceptionHandler.kt
package com.booster.kotlin.shoppingservice.common.exception

import com.booster.kotlin.shoppingservice.common.response.ApiResponse
import com.booster.kotlin.shoppingservice.common.response.ErrorResponse
import org.springframework.http.ResponseEntity
import org.springframework.orm.ObjectOptimisticLockingFailureException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(BusinessException::class)
    fun handleBusinessException(e: BusinessException): ResponseEntity<ApiResponse<Nothing>> {
        val error = ErrorResponse(
            code = e.errorCode.name,
            message = e.message ?: e.errorCode.message
        )
        return ResponseEntity
            .status(e.errorCode.status)
            .body(ApiResponse.fail(error))
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(e: MethodArgumentNotValidException): ResponseEntity<ApiResponse<Nothing>> {
        val message = e.bindingResult.fieldErrors
            .joinToString(", "){ "${it.field}: ${it.defaultMessage}"}
        val error = ErrorResponse(
            code = ErrorCode.INVALID_INPUT.name,
            message = message
        )
        return ResponseEntity
            .status(400)
            .body(ApiResponse.fail(error))
    }

    @ExceptionHandler(ObjectOptimisticLockingFailureException::class)
    fun handleOptimisticLock(e: ObjectOptimisticLockingFailureException): ResponseEntity<ApiResponse<Nothing>> {
        val error = ErrorResponse(
            code = ErrorCode.INSUFFICIENT_STOCK.name,
            message = ErrorCode.INSUFFICIENT_STOCK.message,
        )
        return ResponseEntity.status(409).body(ApiResponse.fail(error))
    }

    @ExceptionHandler(Exception::class)
    fun handleException(e: Exception): ResponseEntity<ApiResponse<Nothing>> {
        val error = ErrorResponse(
            code = ErrorCode.INTERNAL_ERROR.name,
            message = ErrorCode.INTERNAL_ERROR.message
        )
        return ResponseEntity
            .status(500)
            .body(ApiResponse.fail(error))
    }


}